README.md
# react-video-js-player
React wrapper for VideoJS. [Live Demo](https://karan101292.github.io/react-video-js-player/)

## Install
```
npm install --save react-video-js-player
```

## Usage
```javascript
import React, { Component } from 'react';
import VideoPlayer from 'react-video-js-player';

class VideoApp extends Component {
    player = {}
    state = {
        video: {
            src: "http://www.example.com/path/to/video.mp4",
            poster: "http://www.example.com/path/to/video_poster.jpg"
        }
    }

    onPlayerReady(player){
        console.log("Player is ready: ", player);
        this.player = player;
    }

    onVideoPlay(duration){
        console.log("Video played at: ", duration);
    }

    onVideoPause(duration){
        console.log("Video paused at: ", duration);
    }

    onVideoTimeUpdate(duration){
        console.log("Time updated: ", duration);
    }

    onVideoSeeking(duration){
        console.log("Video seeking: ", duration);
    }

    onVideoSeeked(from, to){
        console.log(`Video seeked from ${from} to ${to}`);
    }

    onVideoEnd(){
        console.log("Video ended");
    }

    render() {
        return (
            <div>
                <VideoPlayer
                    controls={true}
                    src={this.state.video.src}
                    poster={this.state.video.poster}
                    width="720"
                    height="420"
                    onReady={this.onPlayerReady.bind(this)}
                    onPlay={this.onVideoPlay.bind(this)}
                    onPause={this.onVideoPause.bind(this)}
                    onTimeUpdate={this.onVideoTimeUpdate.bind(this)}
                    onSeeking={this.onVideoSeeking.bind(this)}
                    onSeeked={this.onVideoSeeked.bind(this)}
                    onEnd={this.onVideoEnd.bind(this)}
                />
            </div>
        );
    }
}
export default VideoApp;
```

### VideoJS APIs support:
> onReady will return <code>videojs</code> instance. Which means you can use all the APIs provided by VideoJS.<br/>[List of VideoJS APIs](https://docs.videojs.com/docs/api/player.html)

### VideoJS plugins support:
> Since most of the VideoJS plugins needs <code>videojs</code> instance to get initialized, it is very easy to integrate any of the available plugins by making use of <code>videojs</code> instance returnd by onReady event.<br/>[List of VideoJS plugins](https://videojs.com/plugins/)

### Available Props:
<table>
  <thead>
    <tr>
      <th>Prop Name</th>
      <th>Prop Type</th>
      <th>Default Value</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>autoplay</td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>Video will start playing automatically if <code>true</code></td>
    </tr>
    <tr>
      <td>bigPlayButton</td>
      <td><code>boolean</code></td>
      <td><code>true</code></td>
      <td>Big play button visibility toggle</td>
    </tr>
    <tr>
      <td>bigPlayButtonCentered</td>
      <td><code>boolean</code></td>
      <td><code>true</code></td>
      <td>Big play button center position toggle</td>
    </tr>
    <tr>
      <td>className</td>
      <td><code>string</code></td>
      <td><code>""</code></td>
      <td>Video player wrapper class. It can be used for custom player skin.</td>
    </tr>
    <tr>
      <td>controls</td>
      <td><code>boolean</code></td>
      <td><code>true</code></td>
      <td>Video player control bar toggle</td>
    </tr>
    <tr>
      <td>height</td>
      <td><code>string | number</code></td>
      <td><code>auto</code></td>
      <td>Video player height</td>
    </tr>
    <tr>
      <td>hideControls</td>
      <td><code>array</code></td>
      <td><code>[]</code></td>
      <td>List of controls to hide. <code>['play','volume','seekbar','timer','playbackrates','fullscreen']</code></td>
    </tr>
    <tr>
      <td>playbackRates</td>
      <td><code>array</code></td>
      <td><code>[0.5, 1, 1.5, 2]</code></td>
      <td>Video speed control</td>
    </tr>
    <tr>
      <td>preload</td>
      <td><code>string</code></td>
      <td><code>auto</code></td>
      <td>video tag preload attribute</td>
    </tr>
    <tr>
      <td>poster</td>
      <td><code>string</code></td>
      <td><code>""</code></td>
      <td>Video poster file path</td>
    </tr>
    <tr>
      <td>src</td>
      <td><code>string</code></td>
      <td><code>""</code></td>
      <td>Video file path</td>
    </tr>
    <tr>
      <td>width</td>
      <td><code>string | number</code></td>
      <td><code>auto</code></td>
      <td>Video player width</td>
    </tr>
    <tr>
      <td>withCredentials</td>
      <td><code>boolean</code></td>
      <td><code>false</code></td>
      <td>When the <code>withCredentials</code> property is set to <code>true</code>, all XHR requests for manifests and segments would have <code>withCredentials</code> set to <code>true</code> as well. This enables storing and passing cookies from the server that the manifests and segments live on. This has some implications on CORS because when set, the <code>Access-Control-Allow-Origin</code> header cannot be set to *, also, the response headers require the addition of <code>Access-Control-Allow-Credentials</code> header which is set to <code>true</code>. See <a href="http://www.html5rocks.com/en/tutorials/cors/" rel="nofollow">html5rocks's article</a> for more info.</td>
    </tr>
  </tbody>
</table>

### Video tracking props:
<table>
  <thead>
    <tr>
      <th>Method Name</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>onEnd</td>
      <td>It will fire when video is finished playing.</td>
    </tr>
    <tr>
      <td>onPause</td>
      <td>It will fire when video is paused. It returns current time of the video</td>
    </tr>
    <tr>
      <td>onPlay</td>
      <td>It will fire when video starts playing anytime. It returns current time of the video</td>
    </tr>
    <tr>
      <td>onReady</td>
      <td>It will fire when video player is ready to be used. It returns <code>videojs</code> instance.</td>
    </tr>
    <tr>
      <td>onSeeked</td>
      <td>It will fire after seeking is done. It returns seek start time and seek end time for the video.</td>
    </tr>
    <tr>
      <td>onSeeking</td>
      <td>It will fire when video is being seeked using seekbar. It returns current time of the video</td>
    </tr>
    <tr>
      <td>onTimeUpdate</td>
      <td>It keeps firing while video is in playing state. It returns current time of the video</td>
    </tr>
  </tbody>
</table>


CHANGELOG.md
# Change Log

| Version | Summary                      |
| ------- | ---------------------------- |
| 1.1.2   | Added option for HLS support |
| 1.1.1   | Initial Release              |


src/index.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Controls from './Controls.json';
import videojs from 'video.js';
import 'video.js/dist/video-js.css';

class VideoPlayer extends Component {
    playerId = `video-player-${(new Date) * 1}`
    player = {};
    componentDidMount() {
        this.init_player(this.props);
        this.init_player_events(this.props);
    }

    componentWillReceiveProps(nextProps) {
        this.set_controls_visibility(this.player, nextProps.hideControls);
        if (this.props.src !== nextProps.src) {
            this.init_player(nextProps);
        }
    }

    componentWillUnmount() {
        if (this.player) this.player.dispose();
    }

    init_player(props) {
        const playerOptions = this.generate_player_options(props);
        this.player = videojs(document.querySelector(`#${this.playerId}`), playerOptions);
        this.player.src(props.src)
        this.player.poster(props.poster)
        this.set_controls_visibility(this.player, props.hideControls);
    }

    generate_player_options(props) {
        const playerOptions = {};
        playerOptions.autoplay = props.autoplay;
        playerOptions.bigPlayButton = props.bigPlayButton;
        playerOptions.controls = props.controls;
        playerOptions.height = props.height;
        playerOptions.preload = props.preload;
        playerOptions.width = props.width;
        playerOptions.withCredentials = props.withCredentials;
        const hidePlaybackRates = props.hidePlaybackRates || props.hideControls.includes('playbackrates');
        if (!hidePlaybackRates) playerOptions.playbackRates = props.playbackRates;
        return playerOptions;
    }

    set_controls_visibility(player, hidden_controls) {
        Object.keys(Controls).map(x => { player.controlBar[Controls[x]].show() })
        hidden_controls.map(x => { player.controlBar[Controls[x]].hide() });
    }

    init_player_events(props) {
        let currentTime = 0;
        let previousTime = 0;
        let position = 0;

        this.player.ready(() => {
            props.onReady(this.player);
            window.player = this.player;
        });
        this.player.on('play', () => {
            props.onPlay(this.player.currentTime());
        });
        this.player.on('pause', () => {
            props.onPause(this.player.currentTime());
        });
        this.player.on('timeupdate', (e) => {
            props.onTimeUpdate(this.player.currentTime());
            previousTime = currentTime;
            currentTime = this.player.currentTime();
            if (previousTime < currentTime) {
                position = previousTime;
                previousTime = currentTime;
            }
        });
        this.player.on('seeking', () => {
            this.player.off('timeupdate', () => { });
            this.player.one('seeked', () => { });
            props.onSeeking(this.player.currentTime());
        });

        this.player.on('seeked', () => {
            let completeTime = Math.floor(this.player.currentTime());
            props.onSeeked(position, completeTime);
        });
        this.player.on('ended', () => {
            props.onEnd();
        });
    }

    render() {
        return (
            <video id={this.playerId} className={`video-js ${this.props.bigPlayButtonCentered ? 'vjs-big-play-centered' : ''} ${this.props.className}`}></video>
        )
    }
}

VideoPlayer.propTypes = {
    autoplay: PropTypes.bool,
    bigPlayButton: PropTypes.bool,
    bigPlayButtonCentered: PropTypes.bool,
    className: PropTypes.string,
    controls: PropTypes.bool,
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    hideControls: PropTypes.arrayOf(PropTypes.string),
    hidePlaybackRates: PropTypes.bool,
    onEnd: PropTypes.func,
    onPause: PropTypes.func,
    onPlay: PropTypes.func,
    onReady: PropTypes.func,
    onSeeked: PropTypes.func,
    onSeeking: PropTypes.func,
    onTimeUpdate: PropTypes.func,
    playbackRates: PropTypes.arrayOf(PropTypes.number),
    poster: PropTypes.string,
    preload: PropTypes.oneOf(['auto', 'none', 'metadata']),
    src: PropTypes.string,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    withCredentials: PropTypes.bool
}

VideoPlayer.defaultProps = {
    autoplay: false,
    bigPlayButton: true,
    bigPlayButtonCentered: true,
    className: "",
    controls: true,
    hideControls: [],
    hidePlaybackRates: false,
    onEnd: () => { },
    onPause: () => { },
    onPlay: () => { },
    onReady: () => { },
    onSeeked: () => { },
    onSeeking: () => { },
    onTimeUpdate: () => { },
    playbackRates: [0.5, 1, 1.5, 2],
    poster: "",
    preload: 'auto',
    src: "",
    withCredentials: false
}

export default VideoPlayer;